examples/BasicOperations/HandlePartialFailure.php

```diff
@@ -0,0 +1,201 @@
+ * proto. Deserializing these to retrieve the error details is may not be immediately obvious at
+ * first, this example shows how to convert Any into <code>GoogleAdsFailure</code>.
+ *
+ * <p>Additionally, this example shows how to produce an error message for a specific failed
+ * operation by looking up the failure details in the <code>GoogleAdsFailure</code> object.
+ */
+class HandlePartialFailure
+{
+    const CUSTOMER_ID = 'INSERT_CUSTOMER_ID_HERE';
+    const CAMPAIGN_ID = 'INSERT_CAMPAIGN_ID_HERE';
+
+    public static function main()
+    {
+        // Either pass the required parameters for this example on the command line, or insert them
+        // into the constants above.
+        $options = (new ArgumentParser())->parseCommandArguments([
+            ArgumentNames::CUSTOMER_ID => GetOpt::REQUIRED_ARGUMENT,
+            ArgumentNames::CAMPAIGN_ID => GetOpt::REQUIRED_ARGUMENT
+        ]);
+
+        // Generate a refreshable OAuth2 credential for authentication.
+        $oAuth2Credential = (new OAuth2TokenBuilder())->fromFile()->build();
+
+        // Construct a Google Ads client configured from a properties file and the
+        // OAuth2 credentials above.
+        $googleAdsClient = (new GoogleAdsClientBuilder())
+            ->fromFile()
+            ->withOAuth2Credential($oAuth2Credential)
+            ->build();
+
+        try {
+            self::runExample(
+                $googleAdsClient,
+                $options[ArgumentNames::CUSTOMER_ID] ?: self::CUSTOMER_ID,
+                $options[ArgumentNames::CAMPAIGN_ID] ?: self::CAMPAIGN_ID
+            );
+        } catch (GoogleAdsException $googleAdsException) {
+            printf(
+                "Request with ID '%s' has failed.%sGoogle Ads failure details:%s",
+                $googleAdsException->getRequestId(),
+                PHP_EOL,
+                PHP_EOL
+            );
+            foreach ($googleAdsException->getGoogleAdsFailure()->getErrors() as $error) {
+                /** @var GoogleAdsError $error */
+                printf(
+                    "\t%s: %s%s",
+                    $error->getErrorCode()->getErrorCode(),
+                    $error->getMessage(),
+                    PHP_EOL
+                );
+            }
+        } catch (ApiException $apiException) {
+            printf(
+                "ApiException was thrown with message '%s'.%s",
+                $apiException->getMessage(),
+                PHP_EOL
+            );
+        }
+    }
+
+    /**
+     * Runs the example.
+     *
+     * @param GoogleAdsClient $googleAdsClient the Google Ads API client
+     * @param int $customerId the client customer ID without hyphens
+     * @param int $campaignId a campaign ID
+     */
+    public static function runExample(GoogleAdsClient $googleAdsClient, $customerId, $campaignId)
+    {
+    
+        $temp_unused_var = new \Google\Ads\GoogleAds\V1\Errors\GoogleAdsFailure();
```

Comment:
Use camelCase + do import of GoogleAdsFailure

Reply:
And this shouldn't be here.
It should be at least very close to when we call mutateAdGroups and you need to explain why we need this.

It'd be better to make your ErrorUtils instantiable and include this line in the constructor instead.
Users wouldn't need to know about this detail.
You can provide this as a constructor parameter to help dependency injection.

Reply:
I don't like much the idea of having the ErrorUtils class instantiable, but I agree that that initialization should not be there. I made an `ErrorUtils::initialize()` function that does this as close as possible to the mutate call but still allows to keep all functions static, as the ErrorUtils class does not in fact have any state.

WDYT?

Reply:
Could you elaborate why you don't like it? Actually, it's used in that way in many of our PHP SOAP library utilities.
Being non-static actually helps when you write test too, as you don't need to worry about inadvertently touch the same parts inadvertently.

BTW, having object instantiation in the initialize() method smells like that we actually need a constructor too.

Reply:
I don't like it because, although it contains an object instantiation, the ErrorUtils does not really have a state itself, so all the methods already in there can be used statically (and, in fact, they are).

The reason why we need to instantiate a GoogleAdsFailure is because of a quirk with the descriptor pool, not because the ErrorUtils class actually uses it or it keeps track of a GoogleAdsFailure object as its state.

As an alternative approach, what if we did the GoogleAdsFailure instantiation inside the `getGoogleAdsFailureFromAny` function in ErrorUtils?

It's the closest place to where it's actually needed and it would be transparent for clients of the ErrorUtils class, without needing to instantiate it or to call `initialize()`.

Reply:
> I don't like it because, although it contains an object instantiation, the ErrorUtils does not really have a state itself, so all the methods already in there can be used statically (and, in fact, they are).

> The reason why we need to instantiate a GoogleAdsFailure is because of a quirk with the descriptor pool, not because the ErrorUtils class actually uses it or it keeps track of a GoogleAdsFailure object as its state.

That's a fair point. I'm fine with that too.

Note that instantiation of GoogleAdsFailure needs to be done only before calling the method that uses partial failure.
So, getGoogleAdsFailureFromAny() cannot be used in this case.
initialize() seems the closest we can get.

Regarding the class structure, can you follow the style I said before?
Other examples than ResourceNames would be [CsvFiles](https://github.com/googleads/googleads-php-lib/blob/master/src/Google/AdsApi/AdManager/Util/v201902/CsvFiles.php) and [AdManagerDates](https://github.com/googleads/googleads-php-lib/blob/master/src/Google/AdsApi/AdManager/Util/v201902/AdManagerDates.php).

After looking at your methods, I think we should divide them into groups:

- PartialFailures contains initialize() (to instantiate GoogleAdsFailure)
- GoogleAdsError contains fromStatus() and fromFailure()
- GoogleAdsFailure contains fromAny() and fromStatus()

Make sure you add tests for them too.

Reply:
Right, I split the ErrorUtils class like you mentioned ad added test coverage.